core/integration/storagetest/batch.go

```diff
@@ -0,0 +1,128 @@
+package storagetest
+
+import (
+	"context"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	spb "github.com/google/keytransparency/core/sequencer/sequencer_go_proto"
+)
+
+// Batcher writes batch definitions to storage.
+type Batcher interface {
+	// WriteBatchSources saves the (low, high] boundaries used for each log in making this revision.
+	WriteBatchSources(ctx context.Context, dirID string, rev int64, meta *spb.MapMetadata) error
+	// ReadBatch returns the batch definitions for a given revision.
+	ReadBatch(ctx context.Context, directoryID string, rev int64) (*spb.MapMetadata, error)
+	// HighestRev returns the highest defined revision number for directoryID.
+	HighestRev(ctx context.Context, directoryID string) (int64, error)
+}
+
+type BatchStorageFactory func(ctx context.Context, t *testing.T) Batcher
+
+type BatchStorageTest func(ctx context.Context, t *testing.T, b Batcher)
+
+// RunBatchStorageTests runs all the batch storage tests against the provided map storage implementation.
+func RunBatchStorageTests(t *testing.T, storageFactory BatchStorageFactory) {
+	ctx := context.Background()
+	b := &BatchTests{}
+	for name, f := range map[string]BatchStorageTest{
+		"TestNotFound":   b.TestNotFound,
+		"TestWriteBatch": b.TestWriteBatch,
+		"TestReadBatch":  b.TestReadBatch,
+		"TestHighestRev": b.TestHighestRev,
+	} {
+		ms := storageFactory(ctx, t)
+		t.Run(name, func(t *testing.T) { f(ctx, t, ms) })
+	}
+}
+
+// BatchTests is a suite of tests to run against
+type BatchTests struct{}
```

Comment:
What's the purpose of this type? Kind of like a "namespace"?

Reply:
The purpose of this type is to allow runtime discovery of `Test*` methods. 
Unlike the compiler, Go cannot do runtime reflection on modules, only concrete types. 
I've keep the runtime reflection bits out of this PR to keep it simple.

Reply:
Is the idea to replace `RunBatchStorageTests` hand-writtent-table-driven function with a code that scans methods using reflection? Maybe add a TODO to clarify that?